staging/src/k8s.io/component-base/zpages/flagz/flagz.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagz
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"text/template"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/apiserver/pkg/endpoints/metrics"
+	"k8s.io/klog/v2"
+
+	cliflag "k8s.io/component-base/cli/flag"
+)
+
+const (
+	flagzTemplate = `
+----------------------------
+title: {{.ComponentName}} flagz
+content_type: reference
+auto_generated: true
+description: flags enabled in {{.ComponentName}}
+----------------------------
+
+`
+)
+
+var (
+	funcMap = template.FuncMap{
+		"ToLower": strings.ToLower,
+	}
+	tmpl          *template.Template
+	flagzRegistry = &Registry{}
+)
+
+type Registry struct {
+	response bytes.Buffer
+	once     sync.Once
+}
+
+type Flagz struct{}
+
+type mux interface {
+	Handle(path string, handler http.Handler)
+}
+
+type templateData struct {
+	ComponentName string
+}
+
+func init() {
+	var err error
+	t := template.New("t").Funcs(funcMap)
+	tmpl, err = t.Parse(flagzTemplate)
+	if err != nil {
+		klog.Errorf("error while parsing gotemplate: %v", err)
+	}
```

Comment:
Even if the feature is disabled, this could theoretically fail because it is in an init function. So the featuregate isn't going to fully serve its purpose of protecting the apiserver from this feature until it becomes stable enough.

But that's just the theory, I wonder if this can really be seen as an issue from a featuregate perspective since I can't see that function fail ever. Either way, it should be safe to move the template initialization to `populateHeader` since it is now protected by a sync.Once.

Reply:
Oh yeah this makes sense. I removed the Init() entirely and put the template initialization under Install() now. Moving it to `populateHeader` started causing one of the tests to fail with `received nil template` errors - perhaps there could be race conditions while 2 tests trying to access the endpoint together.